### ISSUE TITLE
Consider sources.list $(ARCH)

### ISSUE THREAD

**Issue by arkadijs** [AOSC-Dev/oma#14]
Hi, it looks like `oma` doesn't support `$(ARCH)` substitution in `sources.list`. Is it intentional?

**Comment by eatradish** [comment]
There are a couple of scenarios here, if you are adding the architecture from dpkg, `oma refresh` updates are not supported for reading at the moment

If you specify
`[arch=foo,bar]` manually in sources.list, oma is currently supported.

**Comment by eatradish** [comment]
`oma` really doesn't support this right now, `ARCH` support will be added later!

**Comment by MingcongBai** [comment]
> Hi, it looks like `oma` doesn't support `$(ARCH)` substitution in `sources.list`. Is it intentional?

Hi there. We've been looking at this all morning and are a little confused about the feature you requested. Would you care to post your sources.list?

Here's what we tested (which didn't work with APT on AOSC or Ubuntu):

```
deb https://mirrors.bfsu.edu.cn/ubuntu/ noble/binary-amd64 main restricted universe multiverse
```

**Comment by arkadijs** [comment]
```deb` https://nvidia.github.io/libnvidia-container/stable/deb/$(ARCH) /```

It's not hard to manually put amd64 in place yet if `oma` is to fully support APT syntax then this could be useful for some.

Note, ARCH is in the URI, while manual say _suite_.

> suite may also contain a variable, $(ARCH) which expands to the Debian architecture (such as amd64 or armel) used on the system. This permits architecture-independent sources.list files to be used. In general this is only of interest when specifying an exact path; APT will automatically generate a URI with the current architecture otherwise.

https://manpages.debian.org/bullseye/apt/sources.list.5.en.html

**Comment by eatradish** [comment]
Can you try https://github.com/AOSC-Dev/oma/pull/17 ?

**Comment by arkadijs** [comment]
Trying, meanwhile:

1. Building `nettle` requires `pkg-config`. Most devs have it but my install is fresh and it's not mentioned in the oma Readme.

```
--- stderr
Building nettle-sys failed.
Because: Could not run `PKG_CONFIG_ALLOW_SYSTEM_LIBS=1 PKG_CONFIG_ALLOW_SYSTEM_CFLAGS=1 pkg-config --libs --cflags nettle`
The pkg-config command could not be found.
```

2. For that particular [InRelease](https://nvidia.github.io/libnvidia-container/stable/deb/amd64/InRelease) they use SHA512:
```
INFO Refreshing local database ...
ERROR InRelease file is invalid: The SHA256 field is empty.
```

3. While `gcc`, `gcc-runtime`, `rustc` and `glibc` are installed (but no llvm / clang) I have no working `stddef.h` building oma :exploding_head: 
```
  --- stderr
  /usr/include/nettle/nettle-types.h:36:10: fatal error: 'stddef.h' file not found
  thread 'main' panicked at /home/arkadi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/nettle-sys-2.3.0/build.rs:214:39:
  called `Result::unwrap()` on an `Err` value: ClangDiagnostic("/usr/include/nettle/nettle-types.h:36:10: fatal error: 'stddef.h' file not found\n")
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

```
arkadi@x99 [ oma@support-ARCH-in-sources-list ] $ gcc -o a.o -c a.c
arkadi@x99 [ oma@support-ARCH-in-sources-list ] $ cat a.c
#include <stddef.h>
int x(void) { return 1; }
```

```
# find /usr -name stddef.h
/usr/src/linux-headers-6.10.0-aosc-main/include/uapi/linux/stddef.h
/usr/src/linux-headers-6.10.0-aosc-main/include/linux/stddef.h
/usr/include/linux/stddef.h
/usr/lib/gcc/x86_64-aosc-linux-gnu/13.2.0/include/stddef.h
```

```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-aosc-linux-gnu/13.2.0/lto-wrapper
Target: x86_64-aosc-linux-gnu
Configured with: /var/cache/acbs/build/acbs.0lmwgm1t/gcc-13.2.0/configure --build=x86_64-aosc-linux-gnu --prefix=/usr --sysconfdir=/etc --localstatedir=/var --libdir=/usr/lib --bindir=/usr/bin --sbindir=/usr/bin --mandir=/usr/share/man --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://github.com/AOSC-Dev/aosc-os-abbs/issues --enable-shared --enable-threads=posix --with-system-zlib --enable-gnu-indirect-function --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --with-isl=/usr --enable-lto --enable-plugin --enable-install-libiberty --disable-multilib --disable-werror --enable-pie --enable-checking=release --enable-libstdcxx-dual-abi --with-default-libstdcxx-abi=new --enable-default-pie --enable-default-ssp --enable-bootstrap --enable-languages=c,c++,fortran,lto,go,objc,obj-c++,m2,d,ada
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 13.2.0 20230727 (AOSC OS, Core) (GCC) 
```

... which seems to be a separate issue.

**Comment by eatradish** [comment]
> Trying, meanwhile:
> 
>     1. Building `nettle-rs` requires `pkg-config`. Most devs have it but my install is fresh and it's not mentioned in the oma Readme.
> 
> 
> ```
> --- stderr
> Building nettle-sys failed.
> Because: Could not run `PKG_CONFIG_ALLOW_SYSTEM_LIBS=1 PKG_CONFIG_ALLOW_SYSTEM_CFLAGS=1 pkg-config --libs --cflags nettle`
> The pkg-config command could not be found.
> ```
> 
>     2. For that particular [InRelease](https://nvidia.github.io/libnvidia-container/stable/deb/amd64/InRelease) they use SHA512:
> 
> 
> ```
> INFO Refreshing local database ...
> ERROR InRelease file is invalid: The SHA256 field is empty.
> ```
> 
>     3. While `gcc`, `gcc-runtime`, `rustc` and `glibc` are installed (but no llvm / clang) I have no working `stddef.h` nor `limits.h` building oma 🤯
> 
> 
> ```
> --- stderr
> /usr/include/nettle/nettle-types.h:36:10: fatal error: 'stddef.h' file not found
> thread 'main' panicked at /home/arkadi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/nettle-sys-2.3.0/build.rs:214:39:
> called `Result::unwrap()` on an `Err` value: ClangDiagnostic("/usr/include/nettle/nettle-types.h:36:10: fatal error: 'stddef.h' file not found\n")
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```
> 
> ```
> arkadi@x99 [ oma@support-ARCH-in-sources-list ] $ gcc -o a.o -c a.c
> arkadi@x99 [ oma@support-ARCH-in-sources-list ] $ cat a.c
> #include <stddef.h>
> int x(void) { return 1; }
> ```
> 
> ```
> # find /usr -name stddef.h
> /usr/src/linux-headers-6.10.0-aosc-main/include/uapi/linux/stddef.h
> /usr/src/linux-headers-6.10.0-aosc-main/include/linux/stddef.h
> /usr/include/linux/stddef.h
> /usr/lib/gcc/x86_64-aosc-linux-gnu/13.2.0/include/stddef.h
> ```
> 
> ```
> Using built-in specs.
> COLLECT_GCC=gcc
> COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-aosc-linux-gnu/13.2.0/lto-wrapper
> Target: x86_64-aosc-linux-gnu
> Configured with: /var/cache/acbs/build/acbs.0lmwgm1t/gcc-13.2.0/configure --build=x86_64-aosc-linux-gnu --prefix=/usr --sysconfdir=/etc --localstatedir=/var --libdir=/usr/lib --bindir=/usr/bin --sbindir=/usr/bin --mandir=/usr/share/man --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://github.com/AOSC-Dev/aosc-os-abbs/issues --enable-shared --enable-threads=posix --with-system-zlib --enable-gnu-indirect-function --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --with-isl=/usr --enable-lto --enable-plugin --enable-install-libiberty --disable-multilib --disable-werror --enable-pie --enable-checking=release --enable-libstdcxx-dual-abi --with-default-libstdcxx-abi=new --enable-default-pie --enable-default-ssp --enable-bootstrap --enable-languages=c,c++,fortran,lto,go,objc,obj-c++,m2,d,ada
> Thread model: posix
> Supported LTO compression algorithms: zlib zstd
> gcc version 13.2.0 20230727 (AOSC OS, Core) (GCC) 
> ```
> 
> ... which seems to be a separate issue.

1. I'll fix this later in the README
2. this issue is fixed in #17 and seems to work fine
3. compiling oma requires llvm/Clang, which I will explain in the README.
